Replicator/Replicator.cc

```diff
@@ -368,50 +368,110 @@ namespace litecore::repl {
+                        levelSetAt = 4;
                     }
                     DebugAssert(level > kC4Stopped);
                     break;
                 }
             case Connection::kClosing:
                 // Remain active while I wait for the connection to finish closing:
                 logDebug("Connection closing... (activityLevel=busy)waiting to finish");
-                level = kC4Busy;
+                level      = kC4Busy;
+                levelSetAt = 5;
                 break;
             case Connection::kDisconnected:
             case Connection::kClosed:
                 // After connection closes, remain Busy (or Connecting) while I wait for db to
                 // finish writes and for myself to process any pending messages; then go to Stopped.
-                level = Worker::computeActivityLevel();
-                level = max(level, max(_pushStatus.level, _pullStatus.level));
-                if ( level < kC4Busy ) level = kC4Stopped;
-                else if ( currentLevel == kC4Connecting )
-                    level = kC4Connecting;
+                level           = Worker::computeActivityLevel(reason ? &parentReason : nullptr);
+                levelSetAt      = 6;
+                auto childLevel = max(_pushStatus.level, _pullStatus.level);
+                if ( level < childLevel ) {
+                    level      = childLevel;
+                    levelSetAt = 7;
+                }
+                if ( level < kC4Busy ) {
+                    level      = kC4Stopped;
+                    levelSetAt = 8;
+                } else if ( currentLevel == kC4Connecting ) {
+                    level      = kC4Connecting;
+                    levelSetAt = 9;
+                }
                 break;
         }
+        if ( reason ) {
+            switch ( levelSetAt ) {
+                case 0:
+                    *reason = "Connecting";
+                    break;
+                case 1:
+                    *reason = "UnsavedCheckpointer";
+                    break;
+                case 2:
+                    *reason = std::move(parentReason);
+                    break;
+                case 3:
+                    *reason = "pushOrPull";
+                    break;
+                case 4:
+                    *reason = "IdleOneShot";
+                    break;
+                case 5:
+                    *reason = "Closing";
+                    break;
+                case 6:
+                    *reason = "closed:"s + std::move(parentReason);
+                    break;
+                case 7:
+                    *reason = "closed:pushOrPull";
+                    break;
+                case 8:
+                    *reason = "closed";
+                    break;
+                case 9:
+                    *reason = "closed:busyWhenConnecting";
+                    break;
+                default:
+                    DebugAssert(false);
+                    break;
+            }
```

Comment:
It seems like the `reason` return here is just for indicating where the line in the code was executed (It's like I put the breakpoint), and the code added really makes it difficult to understand the original logic. We can do something like "busy as (propery=value, property value)".

Reply:
"It seems like the reason return here is just for indicating where the line in the code was executed (It's like I put the breakpoint)" -- exactly. It's like a breakpoint plus reading off the values in the conditions noting down the actual variable that determines the level.
As for the format of reason, refer to my previous reply.